app/firebase/functions/src/notifications/notification.service.ts

```diff
@@ -0,0 +1,210 @@
+  protected postmark: ServerClient;
+
+  constructor(
+    public db: DBInstance,
+    public notificationsRepo: NotificationsRepository,
+    public postsRepo: PostsRepository,
+    public activityRepo: ActivityRepository,
+    public usersRepo: UsersRepository,
+    config: EmailServiceConfig
+  ) {
+    this.postmark = new postmark.ServerClient(config.apiKey);
+  }
+
+  createNotification(
+    notification: NotificationCreate,
+    manager: TransactionManager
+  ) {
+    return this.notificationsRepo.create(notification, manager);
+  }
+
+  async getFull(
+    userId: string,
+    notificationId: string,
+    manager: TransactionManager
+  ): Promise<NotificationFull> {
+    const notification = await this.notificationsRepo.get(
+      userId,
+      notificationId,
+      manager,
+      true
+    );
+
+    const activity = await this.activityRepo.get(
+      notification.activityId,
+      manager,
+      true
+    );
+
+    return {
+      ...notification,
+      activity,
+    };
+  }
+
+  /** Aggregates all pending notifications of a user and sends them as an email */
+  async notifyUser(userId: string) {
+    await this.db.run(async (manager) => {
+      const user = await this.usersRepo.getUser(userId, manager, true);
+      const settings = user.settings;
+
+      /** get pending notificatations */
+      const pendingIds = await this.notificationsRepo.getUnotifiedOfUser(
+        userId,
+        manager
+      );
+
+      const pendingNotifications = await Promise.all(
+        pendingIds.map((notificationId) =>
+          this.getFull(userId, notificationId, manager)
+        )
+      );
+
+      /** check if user has enabled notififications */
+      if (settings.notificationFreq === NotificationFreq.None) {
+        return;
+      }
+
+      const autopostingOptions = [
+        AutopostOption.AI,
+        AutopostOption.DETERMINISTIC,
+      ];
+
+      if (
+        autopostingOptions.includes(settings.autopost[PLATFORM.Nanopub].value)
+      ) {
+        await this.prepareAndSendDigestAuto(
+          userId,
+          pendingNotifications,
+          manager
+        );
+      } else {
+        await this.prepareAndSendDigestManual(
+          userId,
+          pendingNotifications,
+          manager
+        );
+      }
+
+      await Promise.all(
+        pendingNotifications.map((n) =>
+          this.notificationsRepo.markAsNotified(userId, n.id, manager)
+        )
+      );
+    });
+  }
+
+  async getPostEmailDetails(
+    notification: NotificationFull,
+    types: ActivityType[],
+    manager: TransactionManager
+  ) {
+    if (types.includes(notification.activity.type)) {
+      const post = await this.postsRepo.get(
+        (notification as PostParsedNotification).activity.data.postId,
+        manager,
+        true
+      );
+
+      const postText = PostsHelper.concatenateThread(post);
+
+      return {
+        content: postText,
+        url: getPostUrl(post.id),
+      };
+    } else {
```

Comment:
again here is the assumption that a post parsed activity isn't created for a user with autoposting enabled. I think there would be an edge case if the user first has manual posted set, creates some post which get parsed, then (before the notification is scheduled) changes to autopost. This would mean they will have `pending` PostParsed notifications even though their settings are for autoposting.